res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1136 @@
+            this.connections = [];
+            this.isShifted = false;
+
+            if (options !== undefined && typeof options.key === "string") {
+                this.inKey = options.key;
+                this.outKey = options.key;
+            }
+
+            if (typeof this.unshift === "function") {
+                this.unshift();
+            }
+
+            if (this.outConnect !== false && this.group !== undefined && this.outKey !== undefined) {
+                this.connect();
+                if (this.outTrigger !== false) {
+                    this.trigger();
+                }
+            }
+        }
+        // default attributes
+        // You should probably overwrite at least some of these.
+        inValueScale(value) {
+            // Hack to get exact center of pots to return 0.5
+            if (value > (this.max / 2)) {
+                return (value - 1) / (this.max - 1);
+            } else {
+                return value / (this.max + 1);
+            }
+        }
+        // map input in the XML file, not inValueScale
+        input(channel, control, value, _status, _group) {
+            this.inSetParameter(this.inValueScale(value));
+        }
+        outValueScale(value) { return value * this.max; }
+        output(value, _group, _control) {
+            this.send(this.outValueScale(value));
+        }
+
+        // common functions
+        // In most cases, you should not overwrite these.
+        inGetParameter() {
+            return engine.getParameter(this.group, this.inKey);
+        }
+        inSetParameter(value) {
+            engine.setParameter(this.group, this.inKey, value);
+        }
+        inGetValue() {
+            return engine.getValue(this.group, this.inKey);
+        }
+        inSetValue(value) {
+            engine.setValue(this.group, this.inKey, value);
+        }
+        inToggle() {
+            this.inSetValue(!this.inGetValue());
+        }
+
+        outGetParameter() {
+            return engine.getParameter(this.group, this.outKey);
+        }
+        outSetParameter(value) {
+            engine.setParameter(this.group, this.outKey, value);
+        }
+        outGetValue() {
+            return engine.getValue(this.group, this.outKey);
+        }
+        outSetValue(value) {
+            engine.setValue(this.group, this.outKey, value);
+        }
+        outToggle() {
+            this.outSetValue(!this.outGetValue());
+        }
+
+        connect() {
+            /**
+            Override this method with a custom one to connect multiple Mixxx COs for a single Component.
+            Add the connection objects to the this.connections array so they all get disconnected just
+            by calling this.disconnect(). This can be helpful for multicolor LEDs that show a
+            different color depending on the state of different Mixxx COs. Refer to
+            SamplerButton.connect() and SamplerButton.output() for an example.
+            **/
```

Comment:
Can we use the same commenting style as we do for CPP code, i.e. move the comment out of the body and put it above the function. And maybe use `///`. Not sure if that works with doxygen.

Reply:
What do you think about using JSDoc in JS sources instead?